modules/gapi/src/backends/oak/goakbackend.cpp

```diff
@@ -0,0 +1,616 @@
+        }
+    }
+
+void cv::gimpl::GOAKExecutable::handleNewStream() {
+    // do nothing
+}
+
+void cv::gimpl::GOAKExecutable::handleStopStream() {
+    // do nothing
+}
+
+void cv::gimpl::GOAKExecutable::run(GIslandExecutable::IInput  &in,
+                                    GIslandExecutable::IOutput &out) {
+    const auto in_msg = in.get();
+
+    if (cv::util::holds_alternative<cv::gimpl::EndOfStream>(in_msg)) {
+        out.post(cv::gimpl::EndOfStream{});
+        return;
+    }
+
+    for (size_t i = 0; i < m_in_queues.size(); ++i) {
+        auto q = m_device->getInputQueue(m_in_queues[i].first);
+        q->send(m_in_queues[i].second);
+    }
+
+    for (size_t i = 0; i < m_out_queues.size(); ++i) {
+        auto q = m_out_queues[i];
+        auto oak_frame = q->get<dai::ImgFrame>();
+
+        auto out_arg = out.get(i);
+
+        if (util::holds_alternative<cv::detail::VectorRef>(out_arg)) {
+            cv::util::get<cv::detail::VectorRef>(out_arg).wref<uint8_t>() = oak_frame->getData();
+            // FIXME: should we copy data instead?
+        } else if (util::holds_alternative<cv::MediaFrame*>(out_arg)) {
+            // FIXME: hard-coded NV12
+            *cv::util::get<cv::MediaFrame*>(out_arg) =
+                    cv::MediaFrame::Create<cv::gapi::oak::OAKMediaAdapter>(
+                            cv::Size(static_cast<int>(oak_frame->getWidth()),
+                                     static_cast<int>(oak_frame->getHeight())),
+                            cv::gapi::oak::OAKFrameFormat::NV12,
+                            oak_frame->getData().data(),
+                            oak_frame->getData().data() + static_cast<long>(oak_frame->getData().size() / 3 * 2));
+            // FIXME: should we copy data instead?
+        } else {
+            GAPI_Assert(false && "Unsupported output type");
+        }
+
+        out.meta(out_arg, cv::util::get<cv::GRunArgs>(in_msg)[i].meta);
+        out.post(std::move(out_arg));
+    }
+}
+
+// Built-in kernels for OAK /////////////////////////////////////////////////////
+
+class GOAKBackendImpl final : public cv::gapi::GBackend::Priv {
+    virtual void unpackKernel(ade::Graph            &graph,
+                              const ade::NodeHandle &op_node,
+                              const cv::GKernelImpl &impl) override {
+        OAKGraph gm(graph);
+
+        const auto &kimpl  = cv::util::any_cast<cv::gimpl::GOAKKernel>(impl.opaque);
+        gm.metadata(op_node).set(cv::gimpl::OAKComponent{kimpl});
+    }
+
+    virtual EPtr compile(const ade::Graph &graph,
+                         const cv::GCompileArgs &args,
+                         const std::vector<ade::NodeHandle> &nodes,
+                         const std::vector<cv::gimpl::Data>& ins_data,
+                         const std::vector<cv::gimpl::Data>& outs_data) const override {
+        return EPtr{new cv::gimpl::GOAKExecutable(graph, args, nodes, ins_data, outs_data)};
```

Comment:
Please check here (in metadata) that the graph is compiled for Streaming mode. Throw an error if it is not (as non-streaming mode is not supported by this executable).

Reply:
Done. Question: how could we have non-OAK source in streaming mode, then OAK backend in streaming mode but without camera input? Do we need to support such option at all?

Reply:
that's fairly possible but can be gracefully avoided now.